### ISSUE TITLE
Web-IOU- After requesting money badge in the DM flashes with red 0.00 badge before changing to green badge with requested amount 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#9181]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in with any account
3. Click on Fab menu - Request Money
4. Put amount and click continue
5. Select user and click continue

## Expected Result:
Green badge with the amount requested appears in the DM

## Actual Result:
The badge is switching from 0 (red) to requested amount (green) for a brief moment

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.1.67.0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** applausetester+0901abb@applause.expensifail.com / Feya87Katya
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/170509499-d5a9571c-a6b2-42e0-b3f0-f6dd01b7bf94.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** Applause internal team
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by techievivek** [comment]
Though this can be worked externally I am picking for myself to get comfortable with the codebase.

**Comment by techievivek** [comment]
Will be looking at it today. Thanks

**Comment by techievivek** [comment]
I did get to bottom of this but not sure why I am not able to reproduce this on my machine.
[This](https://github.com/Expensify/App/blob/d2f9d6124a50b8420ce1378a2a3c9422df691c18/src/pages/iou/IOUModal.js#L349) is the part where we are calling the API to create the transaction.
And [this](https://github.com/Expensify/App/blob/d2f9d6124a50b8420ce1378a2a3c9422df691c18/src/libs/actions/IOU.js#L21) is the part where we re-fetch the total IOU amount and update the ONYX.

Re-applying the engineering label so they can confirm if it is reproducible and very much noticable.

**Comment by bfitzexpensify** [comment]
I'm having trouble reproducing this too. I _think_ I got it once, but it was only because I was looking out for it quite intently. I'm not sure this is noticeable enough to be worthwhile to fix.

**Comment by techievivek** [comment]
Closing this for now as per the above comments.

**Comment by kbecciv** [comment]
@techievivek @bfitzexpensify QA team is able to reproduce it with build 1.1.98.0

https://user-images.githubusercontent.com/93399543/189159634-6331328b-e978-4203-bca1-f74a85e113fb.mp4


https://user-images.githubusercontent.com/93399543/189159636-523ced51-5eb8-43db-bfec-4013d8689629.mp4

**Comment by techievivek** [comment]
I think this can be worked externally so passing it to contributors.

**Comment by dylanexpensify** [comment]
internal: https://www.upwork.com/ab/applicants/1569349523573489664/job-details
external: https://www.upwork.com/jobs/~0119fed71c9d8ea05a

**Comment by sobitneupane** [comment]
**Proposal**
**Reason:**
I was able to reproduce the issue only twice. And found that the value of `props.iouReport.ownerEmail` is null at the beginning for some transaction resulting `error={props.session.email !== props.iouReport.ownerEmail}` to be true. Thus, badge remains red for some time.
So, I propose the following change:
https://github.com/Expensify/App/blob/3d76850e2e17baeab5334d4e541e34931e57a62b/src/components/IOUBadge.js#L53-L65
```diff
    return (
+        props.iouReport.ownerEmail && props.session.email ? (
            <Badge
                pressable
                onPress={launchIOUDetailsModal}
                success={props.session.email === props.iouReport.ownerEmail}
                error={props.session.email !== props.iouReport.ownerEmail}
                text={props.numberFormat(
                    props.iouReport.total / 100,
                    {style: 'currency', currency: props.iouReport.currency},
                )}
            />
+        ) : <></>
    );
```

**Comment by iwiznia** [comment]
@parasharrajat can you review the proposal?

**Comment by dylanexpensify** [comment]
Any updates @parasharrajat

**Comment by parasharrajat** [comment]
Seems like simple race conditions issue. Data on optionRowLHN is updated before the real data arrives on IOUBadge. There should not be much delay between these two actions 
https://github.com/Expensify/App/blob/9ff60c3a220c9d05001007b2c0dfe0d954623edb/src/libs/actions/IOU.js#L56-L57.

It only happened once for me. I tested it with one account. A theory could be that.
1. User-created transaction. 
2. Report data updated and hasOutstandingIOU became true.
3. Badge was shown to the user. 
4. But the real data was not present for some reason. 
5. Default props got picked on IOUBadge which resulted in 0 and red color on the badge. 
6. Then real values got pushed via props and all good. 


One strange thing I noticed is that once a badge becomes visible, props are first undefined. 
![image](https://user-images.githubusercontent.com/24370807/193542264-ce58ae2e-7e9f-46b8-8021-c50c8503f6fe.png)



@sobitneupane's [proposal](https://github.com/Expensify/App/issues/9181#issuecomment-1252191160) can be a solution but I don't think it is fixing the real race conditions and adding a workaround. 

IMO, this is a low-priority issue but looks bad from UX perspective.
`

**Comment by dylanexpensify** [comment]
coming back from OOO - reviewing today

**Comment by parasharrajat** [comment]
Waiting for proposals.

**Comment by dylanexpensify** [comment]
Doubling

**Comment by sobitneupane** [comment]
**Proposal:**
As suggested by @parasharrajat, the issue's reason seems to be https://github.com/Expensify/App/issues/9181#issuecomment-1265162385. 

Assuming two merge calls are asynchronous, my proposal is to change order of two merge requests here:
https://github.com/Expensify/App/blob/9ff60c3a220c9d05001007b2c0dfe0d954623edb/src/libs/actions/IOU.js#L56-L57

```
 Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT_IOUS, iouReportsToUpdate); 
 Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, chatReportsToUpdate); 
```

**Comment by sobitneupane** [comment]
> 
> https://github.com/Expensify/App/blob/9ff60c3a220c9d05001007b2c0dfe0d954623edb/src/libs/actions/IOU.js#L56-L57
> 

@parasharrajat Are these merge calls synchronous or asynchronous?

**Comment by parasharrajat** [comment]
Async.

**Comment by tsa321** [comment]
### Proposal 

https://github.com/Expensify/App/blob/adc3ecea49cff208e0003e1a2b37b38ba5be8474/src/components/IOUBadge.js#L39-L50

My solution is to check whether we should display IOUBadge by compare iourepport total with it's default props:

```
return
 ( props.iouReport.total !== defaultProps.iouReport.total && (
        <Badge
            pressable
            onPress={launchIOUDetailsModal}
            success={props.session.email === props.iouReport.ownerEmail}
            error={props.session.email !== props.iouReport.ownerEmail}
            text={props.numberFormat(
                props.iouReport.total / 100,
                {style: 'currency', currency: props.iouReport.currency},
            )}
        />)
)
```

<br />

Another solution is to check IOU report total and compare it with it's default value in `HeaderView`: 
https://github.com/Expensify/App/blob/adc3ecea49cff208e0003e1a2b37b38ba5be8474/src/pages/home/HeaderView.js#L152

This will require `withOnyx` on `iouReport` present in `HeaderView`

**Comment by dylanexpensify** [comment]
No updates - waiting for more proposals. @parasharrajat should we double?

**Comment by parasharrajat** [comment]
No opinion on this.

**Comment by dylanexpensify** [comment]
Going to double

**Comment by dylanexpensify** [comment]
not overdue

**Comment by dylanexpensify** [comment]
reviewing today!

**Comment by dylanexpensify** [comment]
Doubled!

**Comment by ntdiary** [comment]
# Proposal
This issue is easily observed in the Chrome Devtools Performance panel.
The root cause is that reading data from `indexedDB` is much slower than reading data from `cache`.
I think we can use `set` when `originalIOUStatus` is not true.
```diff
diff --git a/src/libs/actions/IOU.js b/src/libs/actions/IOU.js
index ab8aba0d54..fe39b03366 100644
--- a/src/libs/actions/IOU.js
+++ b/src/libs/actions/IOU.js
@@ -147,6 +147,10 @@ function requestMoney(report, amount, currency, recipientEmail, participant, com
         },
     ];

+    if (!originalIOUStatus) {
+        optimisticData[2].onyxMethod = CONST.ONYX.METHOD.SET;
+    }
+
     // Now, let's add the data we need just when we are creating a new chat report
     if (isNewChat) {
         const optimisticCreateAction = ReportUtils.buildOptimisticCreatedReportAction(recipientEmail);
```

**Comment by dylanexpensify** [comment]
@parasharrajat can we get a review of the above proposal?

**Comment by parasharrajat** [comment]
> I want to reproduce it with build 1.1.98.0, but the CreateIOUTransaction api is no longer available, Request Money can't work well.
It is easily reproduced with build 1.2.23-7

Is your proposal still valid? @ntdiary

**Comment by ntdiary** [comment]
> Is your proposal still valid?

Yes, this is a fix for our current code.

<hr />

> https://github.com/Expensify/App/blob/9ff60c3a220c9d05001007b2c0dfe0d954623edb/src/libs/actions/IOU.js#L56-L57

I just wanted to say that this code you discussed before seems to be missing

**Comment by dylanexpensify** [comment]
thoughts on proceeding @parasharrajat?

**Comment by parasharrajat** [comment]
Looks like this issue is no more reproducible. @Expensify/applause Can you please retest?

**Comment by ntdiary** [comment]
This issue is easy to reproduce in my tests, I just don't know if it's necessary to fix:
Version Number: 1.2.24-0

https://user-images.githubusercontent.com/8579651/200328657-42006afc-194e-40ea-b9f4-c0811d4396a4.mp4

![image](https://user-images.githubusercontent.com/8579651/200328849-2cab8790-3427-4bf4-9909-09415fa4c7e8.png)

If we record with Performance panel, it will be easy to observe
1. The red badge lasted for 8 frames, about 880ms

https://user-images.githubusercontent.com/8579651/200333524-7da98cdf-fa28-4e39-a900-aa3ace69deb1.mp4

2. The red badge lasted for 4 frames, about 1880ms !

https://user-images.githubusercontent.com/8579651/200333554-a3ee2fda-9257-49d5-86cc-c746dd6c9008.mp4

**Comment by arielgreen** [comment]
@dylanexpensify can you provide an assessment of what needs to happen to move this one forward? @parasharrajat have you reviewed?

**Comment by dylanexpensify** [comment]
We need:
- @parasharrajat to review proposal as it seems this is still occurring 
- hire contributor, and move forward with a fix

**Comment by ntdiary** [comment]
@parasharrajat hi, any progress on this issue please? Is there anything else I need to do? 🙂

**Comment by dylanexpensify** [comment]
reviewing today!

**Comment by parasharrajat** [comment]
I will be back on this at noon IST.

**Comment by kevinksullivan** [comment]
@parasharrajat do you have an update on the review?

**Comment by dylanexpensify** [comment]
+1 re: @parasharrajat update

**Comment by JmillsExpensify** [comment]
@iwiznia To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by ntdiary** [comment]
@parasharrajat , I'm trying to create a PR. But in testing I also found a problem. I think the server also needs to make some changes.

Here is a chat with IOU history, but the current `hasOutstandingIOU` is `false`.
After we requested money, the server will push two notifications to the browser (type is `onyxApiUpdate`),
the first one will set `reportIOUs_4048917323630499` to `null`,
and the second one will update the `reportIOUs_5603729547185353` data.
If the time difference is large, this will also cause the badge to display red.
**Can we ask the server developer to swap their push order?**

### video
https://user-images.githubusercontent.com/8579651/202689502-06da0394-7360-4f36-8584-9cb4910a11ec.mp4

### websocket:
<img width="1288" alt="websocket" src="https://user-images.githubusercontent.com/8579651/202689602-a790ff4e-9afe-42ff-b9df-889a102bbc84.png">

### log
<img width="1450" alt="log" src="https://user-images.githubusercontent.com/8579651/202689707-b937645b-7ca1-4023-ad87-30de562a6dc9.png">

**Comment by iwiznia** [comment]
I can make that change for you and the proposal looks good, with the only exception that I think you should have the conditional directly [here](https://github.com/Expensify/App/blob/d3eae9a5ea08aff4d6bce132ab5c4bf3f19aa3ba/src/libs/actions/IOU.js#L109-L113) instead of adding it later via `if` and using the `2` index, which could change if people add/remove some actions there.

@dylanexpensify let's hire @ntdiary.
I am sending the server side PR, should be deployed on Tuesday/Wednesday.

**Comment by parasharrajat** [comment]
Thanks, @iwiznia for taking care of the backend and good catch @ntdiary.

**Comment by ntdiary** [comment]
@iwiznia, thanks for your help!  You mean we can do this:
```diff
diff --git a/src/libs/actions/IOU.js b/src/libs/actions/IOU.js
index 438236afde..9fb462f64b 100644
--- a/src/libs/actions/IOU.js
+++ b/src/libs/actions/IOU.js
@@ -113,7 +113,7 @@ function requestMoney(report, amount, currency, recipientEmail, participant, com
             },
         },
         {
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            onyxMethod: originalIOUStatus ? CONST.ONYX.METHOD.MERGE : CONST.ONYX.METHOD.SET,
             key: `${ONYXKEYS.COLLECTION.REPORT_IOUS}${iouReport.reportID}`,
             value: iouReport,
         },
```
<hr />
On the other hand, just to verify the order problem, I made a temporary tweak (this will not be committed):

```diff
diff --git a/src/libs/actions/User.js b/src/libs/actions/User.js
index c3b2556687..a93982937e 100644
--- a/src/libs/actions/User.js
+++ b/src/libs/actions/User.js
@@ -310,10 +310,26 @@ function subscribeToUserEvents() {

     const pusherChannelName = `${CONST.PUSHER.PRIVATE_USER_CHANNEL_PREFIX}${currentUserAccountID}${CONFIG.PUSHER.SUFFIX}`;

+    let deferred;
+    let diffTime;
+
     // Receive any relevant Onyx updates from the server
     PusherUtils.subscribeToPrivateUserChannelEvent(Pusher.TYPE.ONYX_API_UPDATE, currentUserAccountID, (pushJSON) => {
         SequentialQueue.getCurrentRequest().then(() => {
-            Onyx.update(pushJSON);
+            const item = _.find(pushJSON, el => el.key.indexOf('reportIOUs') === 0);
+            if (!item) {
+                Onyx.update(pushJSON);
+            } else if (item.value === null) {
+                deferred = pushJSON;
+                diffTime = Date.now();
+            } else if (pushJSON.length === 4 && deferred) {
+                Onyx.update(pushJSON);
+                diffTime = Date.now() - diffTime;
+                setTimeout(() => {
+                    Onyx.update(deferred);
+                    deferred = null;
+                }, diffTime);
+            }
         });
     });
```
### video

https://user-images.githubusercontent.com/8579651/202741877-b5d68df5-8158-45ec-aac3-cd8b44980c2a.mp4

### log
<img width="1366" alt="fixed-log" src="https://user-images.githubusercontent.com/8579651/202741954-04fc9029-82ce-4186-b953-21eb7b9353d1.png">

We can see that those logs with empty reportID no longer exist.
Finally, I'll keep track of this issue and re-test it after backend changes to make sure it's really fixed. 😄

**Comment by iwiznia** [comment]
> You mean we can do this:

Yep, that way the change stays with the data it belongs to, it's clearer and less opportunities to break it.

> On the other hand, just to verify the order problem, I made a temporary tweak (this will not be committed):

Ah cool! 

> Finally, I'll keep track of this issue and re-test it after backend changes to make sure it's really fixed. 😄

Perfect,

**Comment by iwiznia** [comment]
Waiting on the PR. @ntdiary are you sending it?

**Comment by ntdiary** [comment]
@iwiznia, oh, sorry, just sent it. I tested earlier today and the backend hasn't changed. 😂

**Comment by iwiznia** [comment]
heh yeah, it was deployed to our staging servers today and will be deployed to production some time tomorrow.

**Comment by iwiznia** [comment]
PR is merged

**Comment by JmillsExpensify** [comment]
Woo! Way to help us clear this one out.

**Comment by dylanexpensify** [comment]
@iwiznia are we good to get this paid out now? Or where are we in the countdown

**Comment by dylanexpensify** [comment]
@ntdiary the bounty should be for $4000!

**Comment by dylanexpensify** [comment]
offer sent!

**Comment by dylanexpensify** [comment]
Ah @ntdiary reviewing your comment, you are right irt getting the bonus! I'll make sure when I pay you out to include it! ❤️

**Comment by dylanexpensify** [comment]
@ntdiary payment (w/bonus!) sent!
@parasharrajat sent offer!

**Comment by dylanexpensify** [comment]
payment sent @parasharrajat!
